### DISCUSSION TITLE
When will they implement security function for S7-12/S7-15 PLC ?

### DISCUSSION THREAD

**Discussion by luewh** [gijzelaerr/python-snap7#527] [Polls]
![Snap7 _compatibility](https://github.com/gijzelaerr/python-snap7/assets/85931693/1405827f-449e-44b9-9f66-41ea4209da19)

Maybe the answer is never...
But is there a way to do it by-self ? (even if I know OPC UA could replace Snap7)

**Comment by nikteliy** [comment]
As far as I know, the S1200/S1500 PLCs use the `S7comm+` protocol for security functions. `S7comm+` is an advanced version of `S7comm` with encryption. And precisely because of the encryption, the answer is probably never.

**Reply by luewh** [reply]
OPC UA also has encryption feature, but why it's open source ?
So sad to hear this...

**Comment by spreeker** [comment]
OPC UA is not implemented in the PLC but in the siemen software connecting to the PLC lasttime i checked years ago.  And because of that software in the middle it is never and not nearly as reliable als connecting directly to the plc.  I was thinking the same and could not get my software reliable to work. It took ages to change all the variables I needed to change. Then i found snap7 and wrote this library together with gijs. A truely open source way of connecting to a siemens plc.  And did you know fhe protocal used to talk to the plc S7 is 99.9 identical to an opensource version...but its not...

**Reply by luewh** [reply]
Year, OPC UA is kind of a "third party" unit who read data from PLC and its max rate is 200ms(5 data per second) while Snap7 is 3ms (33 data per second), correct me if I'm wrong.
But, what do you mean exactly by not "reliable" ? It crashes ? It loses data ?

**Comment by lupaulus** [comment]
Having worked for several monthon Siemens products, particularly those communicating in S7. Thanks to this protocol, we can communicate **FAST** between 1 PC/Card and a PLC.  S7 is still the preferred choice for its speed and ease of use compared to OPC (UA/DA). I'm forced to admit that today's PLC networks are increasingly the target of cyber-attacks due to poor network rules. I can only encourage the use of protocols such as OPC UA, but unfortunately these are far too cumbersome and greedy in terms of resources and description. For fast industrial application it is prohibited.

The initiative to launch an S7+ protocol is great (Structure, TIA DataTypes), but as usual Siemens isn't going the open source route, it's starting to work on this connector in proprietary format, as was the case with other protocols long before..
The S7+ connector (1200/1500) is already available on the Siemens Edge platform.

**Reply by luewh** [reply]
Totally agree.

Have you tested Siemens Edge ? It has 1 year free but still require a bank card xD.

**Reply by lupaulus** [reply]
Yeah, we tested it with a Siemens Nanobox IPC227E. It is really complicated for small-medium edge project. Under 100-150 it is not worth of it. Moreover the plateform is in continious develop mode but for the price they want...
The edge Device is based on Docker but without access directly on the docker deamon, you need to use the Edge App Publisher.
The ecosystem is complicated with the Marketplace, the Hub, Industrial Edge Management and the Edge Device.
You can't do anything on the Edge without internet.

Personaly, I selected the Ctrlx Platform based on snapcraft from BoschRexroth that I found more open source with the apps (InfluxDB, Grafana)

**Comment by gijzelaerr** [comment]
can you have a look at this branch? https://github.com/gijzelaerr/python-snap7/pull/603

**Comment by gijzelaerr** [comment]
Update: **S7CommPlus support is now implemented** and available on master (will ship in the next release). The new `s7` package provides a unified client that auto-detects whether the PLC needs legacy S7 or S7CommPlus, including TLS for S7-1200 FW 4.5+ and S7-1500:

```python
from s7 import Client

client = Client()
client.connect("192.168.0.1", 0, 1)  # auto-detects protocol
data = client.db_read(1, 0, 100)
```

Pure Python, no C library, no native dependencies. V2+TLS has been validated in production against real S7-1500 and S7-1200 hardware.
